application/src/main/java/org/togetherjava/tjbot/commands/help/UserBannedDeleteRecentThreadsListener.java

```diff
@@ -0,0 +1,59 @@
+package org.togetherjava.tjbot.commands.help;
+
+import net.dv8tion.jda.api.entities.User;
+import net.dv8tion.jda.api.entities.channel.concrete.ThreadChannel;
+import net.dv8tion.jda.api.events.guild.GuildBanEvent;
+import net.dv8tion.jda.api.hooks.ListenerAdapter;
+import net.dv8tion.jda.api.requests.RestAction;
+import org.jooq.Record1;
+import org.jooq.Result;
+
+import org.togetherjava.tjbot.commands.EventReceiver;
+import org.togetherjava.tjbot.db.Database;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Objects;
+
+import static org.togetherjava.tjbot.db.generated.Tables.HELP_THREADS;
+
+/**
+ * Reacts on users being banned from a guild. It then deletes (not just closes) all their recently
+ * created help threads.
+ * <p>
+ * This for example tackles threads created by scammers, which are otherwise not automatically
+ * cleaned up on ban.
+ */
+public final class UserBannedDeleteRecentThreadsListener extends ListenerAdapter
+        implements EventReceiver {
+    private static final Duration RECENT_THREAD_DURATION = Duration.ofMinutes(30);
+    private final Database database;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param database to find help threads created by the banned user
+     */
+    public UserBannedDeleteRecentThreadsListener(Database database) {
+        this.database = database;
+    }
+
+    @Override
+    public void onGuildBan(GuildBanEvent event) {
+        RestAction.allOf(getRecentHelpThreads(event.getUser()).stream()
+            .map(record -> event.getJDA().getThreadChannelById(record.value1()))
+            .filter(Objects::nonNull)
+            .map(ThreadChannel::delete)
+            .toList()).queue();
```

Comment:
might be wrong, but isnt the behavior now that if a single one of them fails, all of the others fail as well?
and we still get exceptions instead of having a debug-only log.

this is still too fragile and noisy.

Reply:
not it excludes the threads which are null, so ig no exception

Reply:
but u know, delete can fail as well (for various reasons). for example if the thread gets deleted in the meantime or if there are rate limits or if we have discord hicups, ...

we neither want that the other threads fail to get deleted, nor do we want to see exceptions in that case.

this is a cleanup feature and its okay if it fails to cleanup something. it should just log a debug or info but not throw exceptions around

Reply:
> there are rate limits

thats y im combining all the actions and performing all at once to have no rate limit

but the problem is if all the threads are already deleted then only then there will be an exception
```
java.lang.IllegalArgumentException: RestActions may not be empty
```

so ig i can do a normal empty check and if it is null then i'll just won't do anything

Reply:
> if the thread gets deleted in the meantime

it'll get filtered out by `.filter(Objects::nonNull)`

Reply:
![image](https://user-images.githubusercontent.com/73871477/199534262-9acbe35c-3198-4520-ba6c-0470e1538c0b.png)

guess i cant do that without .complete()

Reply:
do u have any alternative method? @Zabuzard

Reply:
ur misunderstanding how RestAction.allOf works.
1. it does not send all of them at once to discord. they are still send individually
2. it does not either "fail all or fail nothing". it will attempt all of them individually and probably abort execution of the rest if just one of them failed, i suppose. u still end up in a mixed state where some worked and some not

what im asking for is: u should send all of them individually and have a proper onFailure listener for them to write a log info or debug level log, instead of sending exceptions all over.

failure of one shouldnt affect the others, hence why u should queue all of them individually.

Reply:
what should be the log message?

Reply:
```java
logger.info("Failed to delete thread {} from banned user.", threadId, failure);
```

Reply:
with some sort of
```java
...stream()
...
forEach(action -> action.queue(any -> {}, failure -> {
  logger.warn(...);
}));
```

Reply:
done